File: YrradiaSDL2/Game/ScenesCore/SceneManager.cs
namespace YrradiaSDL2.Game.ScenesCore
{
    public class SceneManager
    {
        private Dictionary<int, Scene> m_scenes = new Dictionary<int, Scene>();
        private int m_currentScene = 0;

        public void InitializeScenes()
        {
            m_scenes.Add("IntroScene".GetHashCode(), _.introScene);
            m_scenes.Add("MainMenuScene".GetHashCode(), _.mainMenuScene);
            m_scenes.Add("MainScene".GetHashCode(), _.mainScene);
            m_scenes.Add("WorldGenerationScene".GetHashCode(), _.worldGenerationScene);
            GoToScene("IntroScene");
        }

        public void GoToScene(string sceneName)
        {
            m_currentScene = sceneName.GetHashCode();
            _.keyboardInput.Reset();
            m_scenes[m_currentScene].OnEnter();
        }

        public void UpdateCurrentScene()
        {
            if (m_scenes.ContainsKey(m_currentScene))
            {
                m_scenes[m_currentScene].Update();
            }
        }

        public void RenderCurrentScene()
        {
            if (m_scenes.ContainsKey(m_currentScene))
            {
                m_scenes[m_currentScene].Render();
            }
        }
    }
}

File: src/Core/ScenesCore/SceneManager.cpp
#include "SceneManager.hpp"
#include "IScene.hpp"
#include "Theme0/Scenes/Intro/IntroScene.hpp"
#include "Theme0/Scenes/MainMenu/MainMenuScene.hpp"
#include "Theme0/Scenes/WorldGeneration/WorldGenerationScene.hpp"
#include "Theme0/Scenes/Main/MainScene.hpp"

namespace Forradia
{
    SceneManager::SceneManager()
    {
        AddScene("IntroScene", _<IntroScene>());
        AddScene("MainMenuScene", _<MainMenuScene>());
        AddScene("WorldGenerationScene", _<WorldGenerationScene>());
        AddScene("MainScene", _<MainScene>());

        GoToScene("IntroScene");
    }

    void SceneManager::AddScene(std::string_view sceneName,
                                IScene &scene)
    {
        scene.Initialize();

        m_scenes.insert({Hash(sceneName), scene});
    }

    void SceneManager::GoToScene(std::string_view sceneName)
    {
        m_currentScene = Hash(sceneName);

        if (m_scenes.contains(m_currentScene))
        {
            m_scenes.at(m_currentScene).OnEnter();
        }
    }

    void SceneManager::UpdateCurrentScene()
    {
        if (m_scenes.contains(m_currentScene))
        {
            m_scenes.at(m_currentScene).Update();
        }
    }

    void SceneManager::RenderCurrentScene() const
    {
        if (m_scenes.contains(m_currentScene))
        {
            m_scenes.at(m_currentScene).Render();
        }
    }
}
